**Part I: Data Ingestion, Data Manipulation, and Data Export**
- The attached CSV file data.csv has the Home_Score and Away_Score for every event that occurred in an NBA game over the past 5 regular seasons. 

- Please create a new column in the CSV file called Score_Margin using the below formula:
Score_Margin = ABS(Home_Score – Away_Score)
where “ABS” = Absolute Value

- Export the data with the new column Score_Margin to a CSV file called new_data.csv. Your file new_data.csv should contain 6 columns: Season, Game, Event_Num, Home_Score, Away_Score, and Score_Margin.

In [77]:
import pandas as pd

In [78]:
# Read the csv file and print the first five rows

df = pd.read_csv("data.csv")
df

Season  Game  Event_Num  Home_Score  Away_Score
0        2016-17     1          1           0           0
1        2016-17     1          2           0           0
2        2016-17     1          3           0           2
3        2016-17     1          4           0           2
4        2016-17     1          5           0           2
...          ...   ...        ...         ...         ...
2798840  2016-17   438        525         106          95
2798841  2016-17   438        526         106          95
2798842  2016-17   438        528         106          95
2798843  2016-17   439          1           0           0
2798844  2016-17   439          2           0           0

[2798845 rows x 5 columns]

In [79]:
# Create a new column "Score_Margin" with the absolute value of the difference
# between "Home_Score" and "Away_Score"

df['Score_Margin']=(df['Home_Score']-df['Away_Score']).abs()

df

Season  Game  Event_Num  Home_Score  Away_Score  Score_Margin
0        2016-17     1          1           0           0             0
1        2016-17     1          2           0           0             0
2        2016-17     1          3           0           2             2
3        2016-17     1          4           0           2             2
4        2016-17     1          5           0           2             2
...          ...   ...        ...         ...         ...           ...
2798840  2016-17   438        525         106          95            11
2798841  2016-17   438        526         106          95            11
2798842  2016-17   438        528         106          95            11
2798843  2016-17   439          1           0           0             0
2798844  2016-17   439          2           0           0             0

[2798845 rows x 6 columns]

In [80]:
# Creation of a CSV file 

df.to_csv("new_data.csv",index=False)

**Part II: Data Manipulation and Analysis**

- Using the new column Score_Margin, please find the final score margin for each game in the 2020-21 regular season (hint: the final score margin should be calculated using the last row of each game, so there should only be 1 row per game to make the below calculations).

In [85]:
# Select the "Game" and "Score_Margin" columns during the season 2020-21

df = df.loc[df['Season'] == '2020-21']
df = df[['Season','Game','Event_Num','Score_Margin']]
df.sort_values(['Game','Event_Num'])


Season  Game  Event_Num  Score_Margin
2053365  2020-21     1          0            26
2053366  2020-21     1          2             0
2053367  2020-21     1          4             0
2053944  2020-21     1          7             0
2053945  2020-21     1          8             0
...          ...   ...        ...           ...
2636322  2020-21  1080        678             4
2636323  2020-21  1080        679             5
2636324  2020-21  1080        681             5
2636325  2020-21  1080        682             5
2636326  2020-21  1080        683             5

[511994 rows x 4 columns]

There were 1080 games during the season 2020-21.
This number does not seem realistic in my opinion.


In [87]:
# Group the dataframe by game and select only the last row of each game
# Sort the game by ascending value and rectify the index

df = df.groupby('Game')
df = pd.concat([df.tail(1)])
df = df.sort_values(['Game','Event_Num'])
df = df.reset_index(drop=True)
df

Season  Game  Event_Num  Score_Margin
0     2020-21     1        728            26
1     2020-21     2        720             7
2     2020-21     3        684             1
3     2020-21     4        666             4
4     2020-21     5        671            13
...       ...   ...        ...           ...
1075  2020-21  1076        610            16
1076  2020-21  1077        651            28
1077  2020-21  1078        584             2
1078  2020-21  1079        645            12
1079  2020-21  1080        683             5

[1080 rows x 4 columns]

In [88]:
# Use "describe" to get most of the statistic information

df[['Score_Margin']].describe()

Score_Margin
count   1080.000000
mean      11.793519
std        8.948672
min        0.000000
25%        5.000000
50%       10.000000
75%       16.000000
max       57.000000

In [89]:
# Calculate the variance
df[['Score_Margin']].var()

Score_Margin    80.078735
dtype: float64